core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -474,12 +470,23 @@ public boolean process()
                 return false;
             }
 
-            // putIfAbsent will rehash automatically if rehash is needed, unless there isn't enough memory to do so.
-            // Therefore needRehash will not generally return true even if we have just crossed the capacity boundary.
-            while (lastPosition < positionCount && !needRehash()) {
-                int positionInDictionary = block.getId(lastPosition);
-                registerGroupId(dictionary, positionInDictionary);
-                lastPosition++;
+            int remainingPositions = positionCount - lastPosition;
+
+            while (remainingPositions != 0) {
+                int positionCountUntilRehash = maxFill - nextGroupId;
+                if (positionCountUntilRehash == 0) {
+                    if (!tryRehash()) {
+                        return false;
+                    }
+                }
+                int batchSize = min(remainingPositions, BATCH_SIZE);
+
+                for (int i = 0; i < batchSize; i++) {
+                    int positionInDictionary = block.getId(lastPosition);
+                    registerGroupId(dictionary, positionInDictionary);
```

Comment:
actually, I don't understand why it's faster than previous code. We don't take advantage of batching here since `registerGroupId` is not batched. The only benefit seems to be from using `long[] groupIds` in `GetDictionaryGroupIdsWork`